[1/N] Add function to Jsonalize the trace symbol

mzzchy opened the pull request:
Differential Revision: D74404099

mzzchy commented:
@pytorchbot label "topic: not user facing"

Skylion007 commented on torch/csrc/profiler/python/combined_traceback.cpp:
```diff
@@ -115,6 +115,49 @@ struct PythonTraceback : public CapturedTraceback::Python {
+  for (const auto& sc : to_symbolize) {
+    auto it = cached_frames.find(sc);
+    if (it == cached_frames.end()) {
+      cached_frames.insert({sc, unique_frames.size()});
```
Use try_emplace instead? No need to construct a temporary.

mzzchy committed:
[1/N] Add function to Jsonalize the trace symbol (#153158)

Summary:


CUDA Post: https://fb.workplace.com/groups/ai.efficiency.tools.users/permalink/2020094788475989/

# Context
In this diff, we want to enable the on-demand mode of memory snapshot to allow user to trace any remote process via dyno command line.

# Design decision

**How do we send on-demand signal to remote process**
We leverage the dyno-Kineto approach. 
Since dyno is running on all machine in Meta, it can send a request to the remote machine to start the Kineto. 
Kineto will start another thread for memoryProfiler (https://fburl.com/code/dxsmmrok)


**why we use different approach as CUDA**

On CUDA side, we are using pybind to load torch Module and invoke the python api to start/stop the profiling. However, this requires us to compile the whole torch binary in the predictor which is not recommended by runtime(andruwang) 

Thus, we decide to use the CPP api directly to avoid un-necessary dependency

**why the snapshot is saved as json string directly instead of pickle**
Pickle is primarily designed for use with Python and doesn't have well support in cpp. Also, it is hard for user to download the snapshot file and open locally. 
Due to the dependency issue, it is hard to import the gzip/pickle library to decode the data. Thus, let's use JSON for now. I will work on the visualizer to fasten the render and support other format later.

**Plan**: 
* Now, we will encoded file into gz for MTIA ondemand only and update the visualizer to support both type. 
* Update auto-trace and CUDA side to encode in gzip as well
* Fully remove pickle dependency.

Test Plan:
# Remote cogwheel test
Servicelab: https://fburl.com/servicelab/pckux7a3
snapshot file manifold: https://fburl.com/manifold/fnotk18c
snapshot file in pastry: P1805522232

Visualization on D74399684
 {F1977786422}

# Local Predictor Test
url: https://fburl.com/pytorch_memory_visualizer/y06kskkm

 {F1977787329}

Differential Revision: D74404099
